src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -113,20 +113,26 @@ public static ValueTask<ImmutableArray<(Checksum, object)>> ReadDataAsync(PipeRe
                     await localPipe.Writer.CompleteAsync(exception).ConfigureAwait(false);
                     await pipeReader.CompleteAsync(exception).ConfigureAwait(false);
                 }
-            }, cancellationToken).Forget();
+            }, cancellationToken);
 
             // blocking read from the local pipe on the current thread:
             try
             {
                 using var stream = localPipe.Reader.AsStream(leaveOpen: false);
-                return new(ReadData(stream, scopeId, checksums, serializerService, cancellationToken));
+                return ReadData(stream, scopeId, checksums, serializerService, cancellationToken);
             }
             catch (EndOfStreamException)
             {
                 cancellationToken.ThrowIfCancellationRequested();
 
                 throw exception ?? ExceptionUtilities.Unreachable;
             }
+            finally
+            {
+                // Make sure to complete the copy and pipes before returning, otherwise the caller could complete the
+                // reader and/or writer while they are still in use.
```

Comment:
@tmat This failure was caused by `ReadData` knowing _exactly_ how much data it needed, and stopping as soon as that was reached. At that point, this method returned even though the copy operation might not have processed the pipe completions. The caller completed them in the wrong order while the copy was still in progress, and produced an exception indicating something like this:

* The reader cannot be used after completed
* The writer cannot be used after completed

Propagation of the exception had unknown impact on the system, but may have been indirectly responsible for the TaskCanceledException.

Reply:
I see.